server/core/src/password_hash.cpp

```diff
@@ -0,0 +1,311 @@
+            *param++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_SCRYPT_R, &resources);
+            *param++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_SCRYPT_P, &parallelism);
+            // NOLINTEND(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+
+            // Complete construction of scrypt parameters.
+            *param = OSSL_PARAM_construct_end();
+        } // construct_algorithm_parameters
+    }; // class scrypt_kdf
+
+    auto get_kdf(const std::string& _algorithm, const nlohmann::json& _parameters)
+        -> std::unique_ptr<key_derivation_function>
+    {
+        if ("scrypt" == _algorithm) {
+            return std::make_unique<scrypt_kdf>(_parameters);
+        }
+
+        THROW(SYS_NOT_SUPPORTED, fmt::format("{}: Hashing algorithm [{}] not supported.", __func__, _algorithm));
+    } // get_kdf
+} // anonymous namespace
+
+namespace irods
+{
+    auto generate_salt(std::int16_t _salt_length) -> std::string
+    {
+        return irods::generate_random_alphanumeric_string(_salt_length);
+    } // generate_salt
+
+    auto hash_password(const std::string& _password,
+                       const std::string& _salt,
+                       const std::string& _algorithm,
+                       const nlohmann::json& _parameters) -> std::string
+    {
+        // Make sure the provided strings are not empty.
+        if (_password.empty() || _salt.empty()) {
+            THROW(SYS_INVALID_INPUT_PARAM, fmt::format("Cannot derive key from password - password or salt is empty."));
+        }
+
+        // Make sure the length of the provided password does not exceed the allowed maximum.
+        constexpr auto maximum_password_length = 1024 * 1024;
+        if (_password.size() > maximum_password_length) {
+            THROW(PASSWORD_EXCEEDS_MAX_SIZE,
+                  fmt::format("Cannot derive key from password - password length exceeds maximum size."));
+        }
+
+        // Salt is generated by the server and its length is not configurable. Therefore, we do not check it here.
+
+        try {
+            // Using the provided algorithm name and parameters, construct a KDF object.
+            const auto kdf = get_kdf(_algorithm, _parameters);
+
+            // ...and derive a key for the provided password and salt.
+            const auto derived_key = kdf->derive_key(_password, _salt);
+
+            // base64-encode the buffer so that it can be represented as a string. The buffer size is double the
+            // input buffer so as to give more than enough space for the encoded string. Technically, it only needs
+            // a little over (4/3) times the space.
+            std::uint64_t out_len = derived_key.size() * 2;
+            std::vector<unsigned char> out(out_len, 0);
+            auto err = base64_encode(derived_key.data(), derived_key.size(), out.data(), &out_len);
+            if (err < 0) {
+                THROW(err, "base64 encoding of digest failed.");
```

Comment:
Is there a reason these uses of `THROW()` do not capture the name of the function like in other places?